[Two motor tailsitter]Controlling the servo through the output from the mc controller

xdwgood opened the pull request:
I understand that we use the output value from the FW controller to control the aileron in fixed-wing and transmission mode. But in MC mode, the output from the rotor controller should be used to control the aileron.


Now when tailsitter UAV is flying in the rotor state, the pitch / yaw control is through the output of the fw controller instead of the rotor controller output. This PR fixes this problem.


FYI： @RomanBapst

xdwgood commented:
I understand that we use the output value from the FW controller to control the aileron in fixed-wing and conversion mode. But in MC mode, the output from the rotor controller should be used to control the aileron.

xdwgood commented:
@dagar   I saw that the original change came from you, so I want to ask: Why is the pitch / yaw control in rotor mode determined by the FW control, not the MC controller? Thank you  :smiley:

xdwgood commented:
@sfuhrer  Could you take a look?

sfuhrer commented:
> @sfuhrer Could you take a look?

Hi @xdwgood , currently the elevons get always controlled via the FW rate controller, correct. Is is so by design, eg this way every actuator has a dedicated rate controller allocated to it, which we don't change. @RomanBapst could give you further insides why it is done this way. 
What's the reason for this PR, did you encounter some issue with the current implementation?

xdwgood commented:
@sfuhrer   Thank you very much for your reply

>  What's the reason for this PR, did you encounter some issue with the current implementation?


I follow px4 official guidelines to adjust my tailsitter. (https://docs.px4.io/master/en/config_vtol/vtol_quad_configuration.html)   This guide says tuning MC mode using MC controller tuning  `Follow the the respective tuning guides on how to tune multirotors and fixed wings.`    

So I started to adjust the control parameters in mc mode. As a result, I found that changing the PID parameters（eg `MC_PITCHRATE_P, MC_PITCHRATE_I, MC_PITCHRATE_D`） in mc mode had no effect on the attitude control.   

I checked the code and found that we changed the control of the aileron after v1.8.2.
https://github.com/PX4/Firmware/pull/9190   I checked this original PR. Its purpose is to optimize tailsitter control in transmission mode. I don't think we should change the control output logic in mc mode.

xdwgood commented:
@sfuhrer @dagar  Could you give some advice?  thanks

xdwgood commented:
please help

sfuhrer commented:
Hi @xdwgood ,
As far as I understand it the quad tailsitter already uses the MC rate controller to control pitch and yaw, no? (At least the motors, from which I would expect more control authority then from the control surfaces in this case). 
Did you fly this PR on a duo tailsitter as well? And did you see improvements compared to the current master then?
Agree that we should point it out better that for the duo tailsitter one has to tune the FW pitch/roll rate parameter, and also that that's not very intuitive.

xdwgood commented:
@sfuhrer  First, thank you for your reply

Yes, I have used this PR to fly, and the effect is not as good as it is now. I think this is because I did not adjust the pid parameter. Just use the default rotor angular rate control pid parameter.

Even if we use the parameters of fw to adjust the pitch and yaw in the rotor state, I worry that the same parameters are not applicable to the control of pitch and roll in the fixed-wing state.

The control gain in fixed-wing mode may be different from the control gain required in rotor mode, right?

sfuhrer commented:
> @sfuhrer First, thank you for your reply
> 
> Yes, I have used this PR to fly, and the effect is not as good as it is now. I think this is because I did not adjust the pid parameter. Just use the default rotor angular rate control pid parameter.
> 
> Even if we use the parameters of fw to adjust the pitch and yaw in the rotor state, I worry that the same parameters are not applicable to the control of pitch and roll in the fixed-wing state.
> 
> The control gain in fixed-wing mode may be different from the control gain required in rotor mode, right?

Yeah that is indeed a limitation of the current implementation. Do you have an airspeed sensor on your drone? Because that should help in making the transition smoother, as it will automatically adapt the elevon deflection depending on how much airspeed you have.

xdwgood commented:
>  Do you have an airspeed sensor on your drone?

yes，I have.....Your suggestion is that I continue to use FW for rotor attitude control, right?

@sfuhrer  thanks :smiley:

sfuhrer commented:
> > Do you have an airspeed sensor on your drone?
> 
> yes，I have.....Your suggestion is that I continue to use FW for rotor attitude control, right?

Well yeah if it works with the FW rate controller then I would for now use that. 

But I don't want to discourage you from any software changes, it could e.g. be that your proposed change of using the MC rate controller for the control surfaces in hover makes more sense that what there is currently, but to bring in this change in PX4 upstream we would need to be able to compare the flight performance prior/after.

xdwgood commented:
keep

dagar commented:
Closing because the fork disappeared. @xdwgood can you open a new PR with a valid fork if this is still relevant?
